shared/middleware/helpers/dynamo_table.rb

```diff
@@ -0,0 +1,359 @@
+          updated_ip: ip_address,
+          value: value.to_json,
+        },
+        expected: row_id_exists(id),
+      )
+    rescue Aws::DynamoDB::Errors::ConditionalCheckFailedException
+      raise NotFound, "row `#{id}` not found in `#{@table_name}` table"
+    end
+    ensure_column_metadata(value)
+
+    value.merge('id' => id)
+  end
+
+  def ensure_column_metadata(row)
+    ensure_metadata
+    column_list = JSON.parse(metadata['column_list'])
+    row_columns = TableMetadata.generate_column_list([row])
+    new_column_list = column_list.dup.concat(row_columns).uniq
+
+    if column_list != new_column_list
+      set_column_list_metadata(new_column_list)
+    end
+  end
+
+  def items()
+    last_evaluated_key = nil
+
+    [].tap do |results|
+      begin
+        page = db.query(
+          table_name: CDO.dynamo_tables_table,
+          consistent_read: true,
+          key_conditions: {
+            "hash" => {
+              attribute_value_list: [@hash],
+              comparison_operator: "EQ",
+            },
+          },
+          exclusive_start_key: last_evaluated_key,
+        ).first
+
+        page[:items].each do |item|
+          results << item
+        end
+
+        last_evaluated_key = page[:last_evaluated_key]
+      end while last_evaluated_key
+    end
+  end
+
+  def to_a()
+    return items.map { |i| value_from_row(i) }
+  end
+
+  def to_csv()
+    return table_to_csv(to_a, column_order: ['id'])
+  end
+
+  def value_from_row(row)
+    JSON.load(row['value']).merge('id' => row['row_id'].to_i)
+  end
+
+  def self.table_names(channel_id)
+    @dynamo_db ||= Aws::DynamoDB::Client.new
+    last_evaluated_key = nil
+    results = {}
+    begin
+      page = @dynamo_db.query(
+        table_name: CDO.dynamo_tables_table,
+        index_name: CHANNEL_TABLE_NAME_INDEX,
+        key_conditions: {
+          "channel_id" => {
+            attribute_value_list: [channel_id.to_i],
+            comparison_operator: "EQ",
+          },
+        },
+        attributes_to_get: ['table_name'],
+        exclusive_start_key: last_evaluated_key,
+      ).first
+
+      page[:items].each do |item|
+        results[item['table_name']] = true
+      end
+
+      last_evaluated_key = page[:last_evaluated_key]
+    end while last_evaluated_key
+
+    # now same thing for metadata
+    begin
+      page = @dynamo_db.query(
+        table_name: CDO.dynamo_table_metadata_table,
+        index_name: CHANNEL_TABLE_NAME_INDEX,
+        key_conditions: {
+          "channel_id" => {
+            attribute_value_list: [channel_id.to_i],
+            comparison_operator: "EQ",
+          },
+        },
+        attributes_to_get: ['table_name'],
+        exclusive_start_key: last_evaluated_key,
+      ).first
+
+      page[:items].each do |item|
+        results[item['table_name']] = true
+      end
+
+      last_evaluated_key = page[:last_evaluated_key]
+    end while last_evaluated_key
+
+    results.keys
+  end
+
+end
```

Comment:
Unfortunately, by breaking these table implementations out into their own files, we lose the diffs :( If it's helpful, I could create a separate PR just for reviewing purposes that just throws these back into Table.rb to see the diff better. Thoughts?